File: Anul 2/Medii de Programare si Proiectare/mpp-proiect-java-Mothyque/backend/Client/src/main/java/org/example/controller/ServicesRpcProxy.java
package org.example.controller;

import org.example.domain.Client;
import org.example.domain.ClientTicketDTO;
import org.example.domain.Match;
import org.example.domain.User;
import org.example.dto.Request;
import org.example.dto.RequestType;
import org.example.dto.Response;
import org.example.dto.ResponseType;
import org.example.utils.Observer;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.ConnectException;
import java.net.Socket;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingDeque;

public class ServicesRpcProxy implements org.example.networking.Services
{
    private String host;
    private int port;
    private Observer client;
    private ObjectInputStream input;
    private ObjectOutputStream output;
    private Socket connection;
    private BlockingQueue<Response> qresponses;
    private volatile boolean finished;

    public ServicesRpcProxy(String host, int port)
    {
        this.host = host;
        this.port = port;
        qresponses = new LinkedBlockingDeque<>();
    }

    @Override
    public void login(User user, Observer client) throws Exception
    {
        initializeConnection();
        sendRequest(new Request(RequestType.LOGIN, user));
        Response response = readResponse();
        if (response.getType() == ResponseType.OK)
        {
            this.client = client;
        }
        else
        {
            throw new Exception((response.getData().toString()));
        }
    }

    @Override
    public void logout(User user, Observer client) throws Exception
    {
        sendRequest(new Request(RequestType.LOGOUT, user));
        Response response = readResponse();
        closeConnection();
        if (response.getType() == ResponseType.ERROR)
        {
            throw new Exception(response.getData().toString());
        }
    }

    @Override
    public Match[] getAllMatches() throws Exception
    {
        sendRequest(new Request(RequestType.GET_ALL_MATCHES, null));
        Response response = readResponse();
        if (response.getType() == ResponseType.ERROR)
        {
            throw new Exception(response.getData().toString());
        }
        return (Match[]) response.getData();
    }

    @Override
    public void buyTickets(Match match, String clientName, String clientAddress, int numberOfTickets) throws Exception
    {
        Object[] data = new Object[]{match, clientName, clientAddress, numberOfTickets};
        sendRequest(new Request(RequestType.BUY_TICKETS, data));
        Response response = readResponse();
        if (response.getType() == ResponseType.ERROR)
        {
            throw new Exception(response.getData().toString());
        }

    }

    @Override
    public List<ClientTicketDTO> filterTickets(String name, String address) throws Exception
    {
        Object[] data = new Object[]{name, address};
        sendRequest(new Request(RequestType.FILTER_TICKETS, data));
        Response response = readResponse();
        if (response.getType() == ResponseType.ERROR)
        {
            throw new Exception(response.getData().toString());
        }
        return (List<ClientTicketDTO>) response.getData();
    }

    @Override
    public void updateTickets(Match match, Client client, int oldQuantity, int newQuantity) throws Exception {
        Object[] data = new Object[]{match, client, oldQuantity, newQuantity};
        sendRequest(new Request(RequestType.UPDATE_TICKETS, data));
        Response response = readResponse();
        if (response.getType() == ResponseType.ERROR)
        {
            throw new Exception(response.getData().toString());
        }
    }

    @Override
    public void addObserver(Integer userId, Observer client) {}

    void closeConnection()
    {
        try
        {
            if (input != null) input.close();
            if (output != null) output.close();
            if (connection != null) connection.close();
        }
        catch (IOException e)
        {
            e.printStackTrace();
        }
    }

    private void handleUpdate(Response response)
    {
        if (response.getType() == ResponseType.UPDATE)
        {
            client.update();
        }
    }

    private void sendRequest(Request request) throws Exception
    {
        try
        {
            output.writeObject(request);
            output.flush();
        }
        catch (Exception e)
        {
            throw new Exception("Error sending object " + e);
        }
    }

    private Response readResponse() throws Exception
    {
        Response response = null;
        try
        {
            response = qresponses.take();
        }
        catch (InterruptedException e)
        {
            e.printStackTrace();
        }
        return response;
    }

    private void initializeConnection() throws Exception
    {
        try
        {
            connection = new Socket(host, port);
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input = new ObjectInputStream(connection.getInputStream());
            finished = false;
            startReader();
        }
        catch (ConnectException e)
        {
            throw new Exception("Cannot connect to server: " + e.getMessage());
        }
        catch (IOException e)
        {
            throw new Exception("Error initializing connection: " + e.getMessage());
        }
    }

    private void startReader()
    {
        Thread tw = new Thread(new Reader());
        tw.start();
    }

    private class Reader implements Runnable
    {
        public void run()
        {
            while (!finished)
            {
                try
                {
                    Object response = input.readObject();
                    if (isUpdate((Response) response))
                    {
                        handleUpdate((Response) response);
                    }
                    else
                    {
                        try
                        {
                            qresponses.put((Response) response);
                        }
                        catch (InterruptedException e)
                        {
                            e.printStackTrace();
                            Thread.currentThread().interrupt();
                            finished = true;
                        }
                    }
                }
                catch (Exception e)
                {
                    System.out.println("Reading error " + e);
                    finished = true;
                }
            }
        }
    }

    private boolean isUpdate(Response response)
    {
        return response.getType() == ResponseType.UPDATE;
    }

}

File: Anul 2/Medii de Programare si Proiectare/Lab/mpp-proiect-csharp-Mothyque/ProiectMPP/ProiectMPP.Gui/ServicesRpcProxy.cs
using ProiectMPP.Common.Domain;
using ProiectMPP.Common.dto;
using ProiectMPP.Common.Networking;
using ProiectMPP.Common.Utils;
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Net.Sockets;
using System.Threading;
using ClientEntity = ProiectMPP.Common.Domain.Client;

namespace ProiectMPP.Client
{
    public class ServicesRpcProxy : IServices
    {
        private string host;
        private int port;
        private IObserver client;
        private NetworkStream stream;
        private NetworkJsonSerializer serializer;
        private TcpClient connection;
        private BlockingCollection<Response> qresponses;
        private volatile bool finished;

        public ServicesRpcProxy(string host, int port)
        {
            this.host = host;
            this.port = port;
            qresponses = new BlockingCollection<Response>();
        }

        public virtual void Login(User user, IObserver client)
        {
            InitializeConnection();
            SendRequest(new Request(RequestType.LOGIN, user));
            Response response = ReadResponse();
            if (response.Type == ResponseType.OK)
            {
                this.client = client;
            }
            else
            {
                throw new Exception(response.Data.ToString());
            }
        }

        public virtual void Logout(User user, IObserver client)
        {
            SendRequest(new Request(RequestType.LOGOUT, user));
            Response response = ReadResponse();
            CloseConnection();
            if (response.Type == ResponseType.ERROR)
            {
                throw new Exception(response.Data.ToString());
            }
        }

        public virtual Match[] GetAllMatches()
        {
            SendRequest(new Request(RequestType.GET_ALL_MATCHES, null));
            Response response = ReadResponse();
            if (response.Type == ResponseType.ERROR)
            {
                throw new Exception(response.Data.ToString());
            }

            if (response.Data is Match[])
            {
                return (Match[])response.Data;
            }
            else if (response.Data is object[] objArray)
            {
                List<Match> matches = new List<Match>();
                foreach (var obj in objArray)
                {
                    if (obj is Match match)
                    {
                        matches.Add(match);
                    }
                }
                return matches.ToArray();
            }

            throw new Exception("Unexpected response data type: " + response.Data?.GetType().Name);
        }

        public virtual void BuyTickets(Match match, string clientName, string clientAddress, int numberOfTickets)
        {
            object[] data = new object[] { match, clientName, clientAddress, numberOfTickets };
            SendRequest(new Request(RequestType.BUY_TICKETS, data));
            Response response = ReadResponse();
            if (response.Type == ResponseType.ERROR)
            {
                throw new Exception(response.Data.ToString());
            }
        }

        public virtual IList<ClientTicketDTO> FilterTickets(string name, string address)
        {
            object[] data = new object[] { name, address };
            SendRequest(new Request(RequestType.FILTER_TICKETS, data));
            Response response = ReadResponse();
            if (response.Type == ResponseType.ERROR)
            {
                throw new Exception(response.Data.ToString());
            }

            if (response.Data is IList<ClientTicketDTO>)
            {
                return (IList<ClientTicketDTO>)response.Data;
            }
            else if (response.Data is object[] objArray)
            {
                    List<ClientTicketDTO> tickets = new List<ClientTicketDTO>();
                foreach (var obj in objArray)
                {
                    if (obj is ClientTicketDTO dto)
                    {
                        tickets.Add(dto);
                    }
                }
                return tickets;
            }

            throw new Exception("Unexpected response data type: " + response.Data?.GetType().Name);
        }

        public virtual void UpdateTickets(Match match, ProiectMPP.Common.Domain.Client client, int oldQuantity, int newQuantity)
        {
            object[] data = new object[] { match, client, oldQuantity, newQuantity };
            SendRequest(new Request(RequestType.UPDATE_TICKETS, data));
            Response response = ReadResponse();
            if (response.Type == ResponseType.ERROR)
            {
                throw new Exception(response.Data.ToString());
            }
        }

        protected virtual void CloseConnection()
        {
            finished = true;
            try
            {
                stream?.Close();
                connection?.Close();
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        private void HandleUpdate(Response response)
        {
            if (response.Type == ResponseType.UPDATE)
            {
                client.Update();
            }
        }

        private void SendRequest(Request request)
        {
            try
            {
                serializer.SerializeRequest(request);
            }
            catch (Exception e)
            {
                throw new Exception("Error sending request: " + e.Message);
            }
        }

        private void InitializeConnection()
        {
            try
            {
                connection = new TcpClient(host, port);
                stream = connection.GetStream();
                serializer = new NetworkJsonSerializer(stream);
                finished = false;
                StartReader();
            }
            catch (Exception e)
            {
                throw new Exception("Error initializing connection: " + e.Message);
            }
        }

        private void StartReader()
        {
            Thread tw = new Thread(RunReader);
            tw.Start();
        }

        private void RunReader()
        {
            while (!finished)
            {
                try
                {
                    Response res = serializer.DeserializeResponse();
                    if (IsUpdate(res))
                    {
                        HandleUpdate(res);
                    }
                    else
                    {
                        qresponses.Add(res);
                    }
                }
                catch (Exception e)
                {
                    if (!finished)
                    {
                        finished = true;
                        Response errorRes = new Response(ResponseType.ERROR, e.Message);
                        qresponses.Add(errorRes);
                        CloseConnection();
                    }
                }
            }
        }

        private Response ReadResponse()
        {
            Response response = null;
            try
            {
                if (qresponses.TryTake(out response, TimeSpan.FromSeconds(5)))
                {
                    return response;
                }
                else
                {
                    throw new Exception("Timeout waiting for server response.");
                }
            }
            catch (Exception)
            {
                throw;
            }
        }
        private bool IsUpdate(Response response)
        {
            return response.Type == ResponseType.UPDATE;
        }
    }
}
